app/jobs/autoscaling.py

```diff
@@ -102,6 +105,15 @@ def autoscaling_update(config):
         if not is_valid_classification:
             print "WARNING: some instances were classified into multiple states or none. This should never happen."
 
+        if num_discarded > 0 and num_desired > num_available + num_draining:
+            # Case where zombie instances exist.
+            # The "discarded" instances will not in fact get killed, because num_desired is higher than the number
```

Comment:
I guess I would like to understand why a new instance is not launched in this case... Shouldn't `desired` be compared with `available` only? Seems to me that draining should not count as active machines anymore. 
Also, yesterday it looked as if the script was considering the instance still active but busy. Wouldn't this be out of this case? 

Feel free to merge this if you think it mitigates the problem. We can discuss this offline.

Reply:
It doesn't launch a new instance, because from AWS perspective, DesiredCapacity is X and X instances already exist... the AWS autoscaling mechanism itself doesn't know about our draining logic. Conceptually, we would also not want to spin up new instances when there are draining instances because we would rather "rescue" the draining instances into the `available` state to save on init cost. 

However, you have a good point that our scaling mechanism here is a little too complicated.